### DISCUSSION TITLE
Setting non-rectangular boundaries for layout optimization

### DISCUSSION THREAD

**Discussion by Rodrig03** [NatLabRockies/floris#557] [Q&A]
Hello, 

Im running layout optimization tests using LayoutOptimizationScipy from floris.tools.optimization.layout_optimization.layout_optimization_scipy

I am experiencing issues setting irregular wind farm boundaries. It appears that a the most outer "x" and "y" coordinates are taken to create a new rectangle boundary, which is then used for the optimization. 
For example, when setting the following boundaries = [(0, 0), (0, 1600), (1000, 1600), (1600, 1200), (1600, 0), (0, 0)] I get the results shown in the image.
![Figure_4](https://user-images.githubusercontent.com/122991766/213195440-9667ea3f-a32b-4472-b3fc-5eff3c54ae82.png)

Is there any way to set irregular boundaries based on "x", "y" coordinates for layout optimization?

**Comment by cherifM** [comment]
Hi @Rodrig03 

yes you are right, switching to scipy-optimizer is faster but its not working properly, its checking boundary-geometry bounding box, not each  boundary lines, here the boundary is basically polygone. 
I switch back to `PyOptSparse `interface, and everything works, i use `slsqp `, and results are perfect.  This is my solution which I recommend,  I you have to use scipy Optimizer, I did a lot of tests, my conclusions is you have to have really good initial layout, than scipy Optimizer works (bit faster than `PyOptSparse`), but again, `PyOptSparse` is in my opinion is more robust until the **dev team** fix this bug.
hope this help ,

cheers,

cherif

**Comment by bayc** [comment]
@Rodrig03 Thank you for bringing this to our attention. I will look into this shortly and get back with more info.

**Comment by cherifM** [comment]
![Results](https://user-images.githubusercontent.com/7900713/214107120-e98d6c65-6402-4dcf-a92b-2d53798a605b.png)

**Reply by cherifM** [reply]
I did update  stock_examples/18_optimize_layout.py  to use PyOptSparse 
i.e.  use  LayoutOptimizationPyOptSparse interface instead of scipy :  

`
from  floris.tools.optimization.layout_optimization.layout_optimization_pyoptsparse import LayoutOptimizationPyOptSparse 
`


```
layout_opt = LayoutOptimizationPyOptSparse(fi, boundaries ,  freq=freq ,solver = "SLSQP")
sol = layout_opt.optimize()
print(sol )
layout_opt.plot_layout_opt_results()

```

with some changes tin the file :

`.../floris/tools/optimization/layout_optimization/layout_optimization_pyoptsparse.py
`

**Comment by Rodrig03** [comment]
Hello cherifM,

Thank you a lot for looking into this,

The PyOptSparse solution fits the optimized turbine positions into the irregular boundaries. However it seems that you have to have a very good initial layout to use this solution since the turbines are slightly readjusted and not completely repositioned.

Will the option of performing layout optimizations utilizing Scipy be available in a future release?

Also, where can I find the code you updated (stock_examples/18_optimize_layout.py)?
I'm not able to find it in the Git repository.

**Reply by cherifM** [reply]
Hi  @Rodrig03 ,
these are my answers:
1. with  pyoptspare you don't have to have good initial condition, this is issue happend only  with scipy.
2. I don't know when  floris development teams with fix this issue with scipy, im n ot part of this team.
3. the file "stock_examples/18_optimize_layout.py"  is the same file as in "example/18_optimize_layout.py", i did change  it bit like i showed in my previous message, beside this file "floris/tools/optimization/layout_optimization/layout_optimization_pyoptsparse.py". As I explained in my previous message :
(line 93 if i m not wrong) needs to be changed like this  
```
        # Update turbine map with turbince locations
        #self.fi.reinitialize(layout_x = self.x, layout_y = self.y)
        self.fi.reinitialize(layout=(self.x, self.y))
```

and of course you need to install pyoptspare,  please let me know if you need  further help ,

best regards,
cherif
